### PR TITLE
lsETH Integration

### PR DESCRIPTION
**Pull request by cashd**
Branched from #840 

This PR adds Liquid Collective Ether (lsETH) as a yield source. lsETH is a yield bearing token that updates in price everyday by oracles that aggregates and reports the total rewards value accrued by LC's validator set. This integration was heavily influenced by the stETH integration #718 with the following differences. 

- lsETH does not mulDown the sharePrice so certain assertions in the test cases are stricter and check for perfect equality.
- Deposits and withdraws using the base token (ETH) is not supported.

### CODE REVIEW

File: test/instances/lseth/LsETHHyperdrive.t.sol [hunk 1]
```diff
@@ -0,0 +1,901 @@
...
+        vm.startPrank(bob);
+        shortAmount = shortAmount.normalizeToRange(
+            100 * hyperdrive.getPoolConfig().minimumTransactionAmount,
+            HyperdriveUtils.calculateMaxShort(hyperdrive)
+        );
+        RIVER.approve(address(hyperdrive), shortAmount);
+        (uint256 maturityTime, ) = openShort(bob, shortAmount, false);
+
+        // Bob attempts to close the short after the position
+        // duration for ETH. The interest rate range is limited because
```
Comment (cashd) [inline test/instances/lseth/LsETHHyperdrive.t.sol:704]: @jalextowle edge case to look out for here

File: contracts/src/interfaces/lseth/IRiverV1.sol [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+    // /// @return withdrawalEventIds The list of matching withdrawal events, or error codes
+    // function resolveRedeemRequests(
+    //     uint32[] calldata _redeemRequestIds
+    // ) external view returns (int64[] memory withdrawalEventIds);
+
+    // /// @notice Set the daily committable limits
+    // /// @param _dcl The Daily Committable Limits structure
+    // function setDailyCommittableLimits(
+    //     DailyCommittableLimits.DailyCommittableLimitsStruct memory _dcl
+    // ) external;
```
Comment (jalextowle) [inline contracts/src/interfaces/lseth/IRiverV1.sol:220]: Can we delete these lines?

File: contracts/src/interfaces/lseth/IRiverV1.sol [hunk 3]
```diff
@@ -0,0 +1,279 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity 0.8.20;
+
+import { ISharesManagerV1 } from "./ISharesManagerV1.sol";
+
+/// @title River Interface (v1)
+/// @author Kiln
+/// @notice The main system interface
+interface IRiverV1 is ISharesManagerV1 {
```
Comment (jalextowle) [inline contracts/src/interfaces/lseth/IRiverV1.sol:9]: We should consider trimming down the interfaces for integrations. We can remove 90% of these functions since most are unused in the integration. Same for the other integrations.

File: contracts/src/interfaces/lseth/ISharesManagerV1.sol [hunk 4]
```diff
@@ -0,0 +1,132 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity 0.8.20;
+
+import "openzeppelin-contracts/contracts/token/ERC20/IERC20.sol";
+
+/// @title Shares Manager Interface (v1)
+/// @author Kiln
+/// @notice This interface exposes methods to handle the shares of the depositor and the ERC20 interface
+interface ISharesManagerV1 is IERC20 {
```
Comment (jalextowle) [inline contracts/src/interfaces/lseth/ISharesManagerV1.sol:9]: Same here, we don't need most of these functions for the integration or tests.

File: contracts/src/instances/lseth/LsETHBase.sol [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+
+    /// @notice Instantiates the lsETH Hyperdrive base contract.
+    /// @param __river The lsETH contract.
+    constructor(IRiverV1 __river) {
+        _river = __river;
+    }
+
+    /// Yield Source ///
+
+    /// @dev Accepts a deposit from the user in base.
```
Comment (jalextowle) [inline contracts/src/instances/lseth/LsETHBase.sol:33]: Let's add a sentence about how this option is unsupported for this yield source.

File: contracts/src/instances/lseth/LsETHBase.sol [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+    /// @return sharesMinted The shares that were minted in the deposit.
+    /// @return refund The amount of ETH to refund. This should be zero for
+    ///         yield sources that don't accept ETH.
```
Comment (jalextowle) [inline contracts/src/instances/lseth/LsETHBase.sol:36]: We don't have the named returns `sharesMinted` or `refund` in the function signature, so we'll want to remove those names from the NatSpec.

Comment (jalextowle) [reply contracts/src/instances/lseth/LsETHBase.sol:36]: We can remove these lines of the comment since the function just reverts.

File: contracts/src/instances/lseth/LsETHBase.sol [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+        address _destination,
+        bytes calldata // unused
+    ) internal override {
+        // Transfer the lsETH shares to the destination.
+        _river.transfer(_destination, _shareAmount);
+    }
+
+    /// @dev We override the message value check since this integration is
+    ///      payable.
+    function _checkMessageValue() internal pure override {}
```
Comment (jalextowle) [inline contracts/src/instances/lseth/LsETHBase.sol:81]: The integration shouldn't be payable since we don't accept the base option. Let's lock this down.

File: test/instances/lseth/LsETHHyperdrive.t.sol [hunk 8]
```diff
@@ -0,0 +1,901 @@
...
+import { IRiverV1 } from "contracts/src/interfaces/lseth/IRiverV1.sol";
+import { Lib } from "test/utils/Lib.sol";
+import { LsETHHyperdriveCoreDeployer } from "contracts/src/deployers/lseth/LsETHHyperdriveCoreDeployer.sol";
+import { LsETHHyperdriveDeployerCoordinator } from "contracts/src/deployers/lseth/LsETHHyperdriveDeployerCoordinator.sol";
+import { LsETHTarget0Deployer } from "contracts/src/deployers/lseth/LsETHTarget0Deployer.sol";
+import { LsETHTarget1Deployer } from "contracts/src/deployers/lseth/LsETHTarget1Deployer.sol";
+import { LsETHTarget2Deployer } from "contracts/src/deployers/lseth/LsETHTarget2Deployer.sol";
+import { LsETHTarget3Deployer } from "contracts/src/deployers/lseth/LsETHTarget3Deployer.sol";
+import { LsETHTarget4Deployer } from "contracts/src/deployers/lseth/LsETHTarget4Deployer.sol";
+import { stdStorage, StdStorage } from "forge-std/Test.sol";
```
Comment (jalextowle) [inline test/instances/lseth/LsETHHyperdrive.t.sol:24]: Nit: External libraries should be at the top of the import statements.

File: test/instances/lseth/LsETHHyperdrive.t.sol [hunk 9]
```diff
@@ -0,0 +1,901 @@
...
+import { IHyperdrive } from "contracts/src/interfaces/IHyperdrive.sol";
+import { IRiverV1 } from "contracts/src/interfaces/lseth/IRiverV1.sol";
+import { Lib } from "test/utils/Lib.sol";
+import { LsETHHyperdriveCoreDeployer } from "contracts/src/deployers/lseth/LsETHHyperdriveCoreDeployer.sol";
+import { LsETHHyperdriveDeployerCoordinator } from "contracts/src/deployers/lseth/LsETHHyperdriveDeployerCoordinator.sol";
+import { LsETHTarget0Deployer } from "contracts/src/deployers/lseth/LsETHTarget0Deployer.sol";
+import { LsETHTarget1Deployer } from "contracts/src/deployers/lseth/LsETHTarget1Deployer.sol";
+import { LsETHTarget2Deployer } from "contracts/src/deployers/lseth/LsETHTarget2Deployer.sol";
+import { LsETHTarget3Deployer } from "contracts/src/deployers/lseth/LsETHTarget3Deployer.sol";
+import { LsETHTarget4Deployer } from "contracts/src/deployers/lseth/LsETHTarget4Deployer.sol";
```
Comment (jalextowle) [inline test/instances/lseth/LsETHHyperdrive.t.sol:23]: It doesn't look like these are in alphabetical order. Let's fix that. `test/` should be below `contracts`

File: test/instances/lseth/LsETHHyperdrive.t.sol [hunk 10]
```diff
@@ -0,0 +1,901 @@
...
+import { stdStorage, StdStorage } from "forge-std/Test.sol";
+import "forge-std/console.sol";
```
Comment (jalextowle) [inline test/instances/lseth/LsETHHyperdrive.t.sol:25]: Nit: Remove this.

```suggestion
```

File: test/instances/lseth/LsETHHyperdrive.t.sol [hunk 11]
```diff
@@ -0,0 +1,901 @@
...
+        super.setUp();
+
+        // Fund the test accounts with lsETH and ETH.
+        address[] memory accounts = new address[](3);
+        accounts[0] = alice;
+        accounts[1] = bob;
+        accounts[2] = celine;
+        fundAccounts(
+            address(hyperdrive),
+            IERC20(0x8c1BEd5b9a0928467c9B1341Da1D7BD5e10b6549),
```
Comment (jalextowle) [inline test/instances/lseth/LsETHHyperdrive.t.sol:58]: Can't we just use the RIVER constant here?

File: test/instances/lseth/LsETHHyperdrive.t.sol [hunk 12]
```diff
@@ -0,0 +1,901 @@
...
+
+        // Start recording event logs.
+        vm.recordLogs();
+    }
+
+    /// Deploy and Initialize ///
+
+    function test__lseth__deployAndInitialize() external {
+        // Deploy and Initialize the lsETH hyperdrive instance. Excess ether is
+        // sent, and should be returned to the sender.
```
Comment (jalextowle) [inline test/instances/lseth/LsETHHyperdrive.t.sol:234]: Similar to the comment I left in the RETH PR, this comment doesn't make sense since we aren't sending ETH.

File: test/instances/lseth/LsETHHyperdrive.t.sol [hunk 13]
```diff
@@ -0,0 +1,901 @@
...
+
+        // Ensure that Bob received the correct amount of LP tokens. He should
+        // receive LP shares totaling the amount of shares that he contributed
+        // minus the shares set aside for the minimum share reserves and the
+        // zero address's initial LP contribution.
+        assertApproxEqAbs(
+            hyperdrive.balanceOf(AssetId._LP_ASSET_ID, bob),
+            contribution - 2 * hyperdrive.getPoolConfig().minimumShareReserves,
+            1e5
+        );
```
Comment (jalextowle) [inline test/instances/lseth/LsETHHyperdrive.t.sol:338]: We should be able to make this an `assertEq` since the LP shares are denominated in shares.

File: test/instances/lseth/LsETHHyperdrive.t.sol [hunk 14]
```diff
@@ -0,0 +1,901 @@
...
+            bobBalancesBefore,
+            hyperdriveBalancesBefore
+        );
+    }
+
+    function test_open_long_refunds() external {
+        vm.startPrank(bob);
+
+        // Ensure that the refund fails when Bob sends excess ETH
+        // when opening a long with "asBase" set to true.
```
Comment (jalextowle) [inline test/instances/lseth/LsETHHyperdrive.t.sol:450]: Nit: It's not really the refund that is failing. This will fail before a refund is attempted.

File: test/instances/lseth/LsETHHyperdrive.t.sol [hunk 15]
```diff
@@ -0,0 +1,901 @@
...
+            bob,
+            sharesPaid,
+            false
+        );
+
+        // The term passes and some interest accrues.
+        variableRate = variableRate.normalizeToRange(0, 2.5e18);
+        advanceTime(POSITION_DURATION, variableRate);
+
+        vm.expectRevert(IHyperdrive.UnsupportedToken.selector);
```
Comment (jalextowle) [inline test/instances/lseth/LsETHHyperdrive.t.sol:515]: Nit: Let's add a comment explaining this.

File: test/instances/lseth/LsETHHyperdrive.t.sol [hunk 16]
```diff
@@ -0,0 +1,901 @@
...
+            100 * hyperdrive.getPoolConfig().minimumTransactionAmount,
+            HyperdriveUtils.calculateMaxShort(hyperdrive)
+        );
+        RIVER.approve(address(hyperdrive), shortAmount);
+        (uint256 maturityTime, ) = openShort(bob, shortAmount, false);
+
+        // Bob attempts to close the short after the position
+        // duration for ETH. The interest rate range is limited because
+        // the test case not fail when there is zero accrued interest.
+        variableRate = variableRate.normalizeToRange(1e18, 2.5e18);
```
Comment (jalextowle) [inline test/instances/lseth/LsETHHyperdrive.t.sol:706]: This is probably higher than it needs to be. I ran into the same edge case in another integration and was able to use `0.01e18` as the lower bound: https://github.com/delvtech/hyperdrive/blob/ee0d8cf64bef6de897bfc7532a2684ed5534d7f5/test/instances/steth/StETHHyperdrive.t.sol#L687

File: test/instances/lseth/LsETHHyperdrive.t.sol [hunk 17]
```diff
@@ -0,0 +1,901 @@
...
+        uint256 amount,
+        bool asBase,
+        uint256 totalSharesBefore,
+        AccountBalances memory traderBalancesBefore,
+        AccountBalances memory hyperdriveBalancesBefore
+    ) internal {
+        if (asBase) {
+            revert IHyperdrive.UnsupportedToken();
+        }
+        // Ensure that the ether balances were updated correctly.
```
Comment (jalextowle) [inline test/instances/lseth/LsETHHyperdrive.t.sol:791]: Nit: Newline

```suggestion

        // Ensure that the ether balances were updated correctly.
```

**Review by jalextowle — COMMENTED** [review]
This looks great, nice job! I left some comments, and I'll take another quick look once those are addressed.

File: contracts/src/deployers/lseth/LsETHHyperdriveDeployerCoordinator.sol [hunk 18]
```diff
@@ -0,0 +1,120 @@
...
+                .InvalidMinimumTransactionAmount();
+        }
+    }
+
+    /// @dev Gets the initial vault share price of the Hyperdrive pool.
+    /// @return The initial vault share price of the Hyperdrive pool.
+    function _getInitialVaultSharePrice(
+        bytes memory // unused extra data
+    ) internal view override returns (uint256) {
+        // Return LsETH's current vault share price.
```
Comment (jalextowle) [inline contracts/src/deployers/lseth/LsETHHyperdriveDeployerCoordinator.sol:117]: Nit: 

```suggestion
        // Return lsETH's current vault share price.
```

File: contracts/src/interfaces/lseth/IRiverV1.sol [hunk 19]
```diff
@@ -0,0 +1,21 @@
...
+    /// @param requested The amount that was requested
+    /// @param received The amount that was received
+    error InvalidPulledClFundsAmount(uint256 requested, uint256 received);
+
+    /// @notice The computed amount of shares to mint is 0
+    error ZeroMintedShares();
+
+    /// @notice The access was denied
+    /// @param account The account that was denied
+    error Denied(address account);
```
Comment (jalextowle) [inline contracts/src/interfaces/lseth/IRiverV1.sol:20]: We still don't use any of these errors, right? Why keep them?

File: contracts/src/interfaces/lseth/ISharesManagerV1.sol [hunk 20]
```diff
@@ -0,0 +1,106 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity 0.8.20;
+
+import { IERC20 } from "../IERC20.sol";
+
+/// @title Shares Manager Interface (v1)
+/// @author Kiln
+/// @notice This interface exposes methods to handle the shares of the depositor and the ERC20 interface
+interface ISharesManagerV1 is IERC20 {
```
Comment (jalextowle) [inline contracts/src/interfaces/lseth/ISharesManagerV1.sol:9]: Same here, it seems like most of the interface is unused.

**Review by jalextowle — APPROVED** [review]
This is looking great @cashd! Really nice job. Just a few nits, but nothing major.
